flink-runtime/src/test/java/org/apache/flink/runtime/jobmaster/utils/JobMasterBuilder.java

```diff
@@ -0,0 +1,211 @@
+		testingHighAvailabilityServices.setResourceManagerLeaderRetriever(rmLeaderRetrievalService);
+
+		this.highAvailabilityServices = testingHighAvailabilityServices;
+		this.jobGraph = jobGraph;
+		this.rpcService = rpcService;
+	}
+
+	public JobMasterBuilder withConfiguration(Configuration configuration) {
+		this.configuration = configuration;
+		return this;
+	}
+
+	public JobMasterBuilder withHighAvailabilityServices(HighAvailabilityServices highAvailabilityServices) {
+		this.highAvailabilityServices = highAvailabilityServices;
+		return this;
+	}
+
+	public JobMasterBuilder withJobManagerSharedServices(JobManagerSharedServices jobManagerSharedServices) {
+		this.jobManagerSharedServices = jobManagerSharedServices;
+		return this;
+	}
+
+	public JobMasterBuilder withHeartbeatServices(HeartbeatServices heartbeatServices) {
+		this.heartbeatServices = heartbeatServices;
+		return this;
+	}
+
+	public JobMasterBuilder withSlotPoolFactory(SlotPoolFactory slotPoolFactory) {
+		this.slotPoolFactory = slotPoolFactory;
+		return this;
+	}
+
+	public JobMasterBuilder withFatalErrorHandler(FatalErrorHandler fatalErrorHandler) {
+		this.fatalErrorHandler = fatalErrorHandler;
+		return this;
+	}
+
+	public JobMasterBuilder withSchedulerFactory(SchedulerFactory schedulerFactory) {
+		this.schedulerFactory = schedulerFactory;
+		return this;
+	}
+
+	public JobMasterBuilder withSchedulerNGFactory(SchedulerNGFactory schedulerNGFactory) {
+		this.schedulerNGFactory = schedulerNGFactory;
+		return this;
+	}
+
+	public JobMasterBuilder withOnCompletionActions(OnCompletionActions onCompletionActions) {
+		this.onCompletionActions = onCompletionActions;
+		return this;
+	}
+
+	public JobMasterBuilder withResourceId(ResourceID resourceId) {
+		this.jmResourceId = resourceId;
+		return this;
+	}
+
+	public JobMasterBuilder withShuffleMaster(ShuffleMaster<?> shuffleMaster) {
+		this.shuffleMaster = shuffleMaster;
+		return this;
+	}
+
+	public JobMasterBuilder withPartitionTrackerFactory(PartitionTrackerFactory partitionTrackerFactory) {
+		this.partitionTrackerFactory = partitionTrackerFactory;
+		return this;
+	}
+
+	public JobMaster createJobMaster() throws Exception {
+		final JobMasterConfiguration jobMasterConfiguration = JobMasterConfiguration.fromConfiguration(configuration);
+
+		return new JobMaster(
+			rpcService,
+			jobMasterConfiguration,
+			jmResourceId,
+			jobGraph,
+			highAvailabilityServices,
+			slotPoolFactory != null ? slotPoolFactory : DefaultSlotPoolFactory.fromConfiguration(configuration),
+			schedulerFactory != null ? schedulerFactory : DefaultSchedulerFactory.fromConfiguration(configuration),
+			jobManagerSharedServices,
+			heartbeatServices,
+			UnregisteredJobManagerJobMetricGroupFactory.INSTANCE,
+			onCompletionActions,
+			fatalErrorHandler,
+			JobMasterBuilder.class.getClassLoader(),
+			schedulerNGFactory != null ? schedulerNGFactory : new LegacySchedulerFactory(jobManagerSharedServices.getRestartStrategyFactory()),
```

Comment:
Why don't we initialize `slotPoolFactory`, `schedulerFactory` and `schedulerNGFactory` with the default factories? Then we would not have to check whether they are `null` or not.

Reply:
This is done to prevent unexpected behaviors when a user sets a configuration/sharedServices but no factory.
In this case the default factories would not rely on the given configuration/sharedServices, but a user would likely expect that. This could be tricky to track down.

We also can't recreate the factories once the configuration is set because we may be overwriting an existing factory that was explicitly set. (for which we'd need additional flags, voiding the original purpose).

Reply:
True. Thanks for the clarification.